rpc/providers/proxy.ts

```diff
@@ -1,46 +1,28 @@
 import { Config } from "../../config/mod.ts";
-import { deadline, deferred } from "../../deps/std/async.ts";
 import { ErrorCtor } from "../../util/mod.ts";
 import * as B from "../Base.ts";
 import { ClientHooks, ParseRawIngressMessageError } from "../common.ts";
+import { WsContainer } from "./ws_container.ts";
 
 export type ProxyClientHooks<Config_ extends Config<string>> = ClientHooks<Config_, Event>;
 
 export function proxyClient<Config_ extends Config<string>>(
   config: Config_,
-  hooks?: ProxyClientHooks<Config_>,
 ): Promise<ProxyClient<Config_> | FailedToOpenConnectionError> {
-  const ws = new WebSocket(config.discoveryValue);
-  const client = new ProxyClient(ws, hooks);
-  ws.addEventListener("error", client.onError);
-  ws.addEventListener("message", client.onMessage);
-  const pending = deferred<ProxyClient<Config_> | FailedToOpenConnectionError>();
-  if (ws.readyState === WebSocket.CONNECTING) {
-    const onOpenError = (e: any) => {
-      console.log({ log: e });
-      clearListeners();
-      pending.resolve(new FailedToOpenConnectionError());
-    };
-    const onOpen = () => {
-      clearListeners();
-      pending.resolve(client);
-    };
-    const clearListeners = () => {
-      ws.removeEventListener("error", onOpenError);
-      ws.removeEventListener("open", onOpen);
-    };
-    ws.addEventListener("error", onOpenError);
-    ws.addEventListener("open", onOpen);
-  } else {
-    pending.resolve(client);
-  }
-  return pending;
+  const ws = new WsContainer({
+    webSocketFactory: () => new WebSocket(config.discoveryValue),
+  });
+
+  return new Promise((resolve) => {
+    ws.once("open", () => resolve(new ProxyClient(ws)));
+    ws.once("error", () => resolve(new FailedToOpenConnectionError()));
```

Comment:
I'm still not very convinced with this.
One or the other will stay attached.
Perhaps something like this would be better https://deno.land/std@0.156.0/node/events.ts?s=once

or 
```
ws.once(["open", "error"], (event) => ...);
```